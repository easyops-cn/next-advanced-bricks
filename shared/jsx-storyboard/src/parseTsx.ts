import { parse, type ParseResult } from "@babel/parser";
import * as t from "@babel/types";
import type {
  Component,
  ConstructResult,
  DataSource,
  ParseError,
  ParseJsxOptions,
  Variable,
} from "./interfaces.js";
import { constructJsValue } from "./tsx-constructors/values.js";
import { constructTsxView } from "./tsx-constructors/view.js";
import { parseTsxCallApi } from "./tsx-constructors/api.js";
import { replaceVariables } from "./tsx-constructors/replaceVariables.js";

export function parseTsx(source: string, options?: ParseJsxOptions) {
  const dataSources: DataSource[] = [];
  const variables: Variable[] = [];
  const components: Component[] = [];
  const errors: ParseError[] = [];
  let title: string | undefined;
  const componentsMap = new Map<string, Component>();
  const contexts: string[] = [];
  const contracts = new Set<string>();
  const result: ConstructResult = {
    source,
    title,
    dataSources,
    variables,
    components,
    componentsMap,
    errors,
    contexts,
    contracts,
  };

  let ast: ParseResult<t.File> | undefined;
  try {
    ast = parse(source, {
      plugins: ["jsx", "typescript"],
      sourceType: "module",
      errorRecovery: true,
    });
  } catch (error) {
    errors.push({
      message: `Failed to parse TSX: ${error}`,
      node: null,
      severity: "fatal",
    });
  }

  if (ast?.errors?.length) {
    for (const error of ast.errors) {
      errors.push({
        message: `${error.code}: ${error.reasonCode}`,
        node: null,
        severity: "error",
      });
    }
  }

  function parseDataSourceCall(
    call: t.Expression,
    name: string,
    transformArgs?: t.CallExpression["arguments"]
  ) {
    const payload = parseTsxCallApi(call, result, options);
    if (!payload) {
      return null;
    }
    contracts.add(payload.api);
    const dataSource: DataSource = {
      name,
      ...payload,
    };

    if (transformArgs) {
      if (transformArgs.length !== 1) {
        errors.push({
          message: `"callApi().then()" expects exactly 1 argument, but got ${transformArgs.length}`,
          node: call,
          severity: "error",
        });
        return false;
      }
      const transform = transformArgs[0];
      if (!t.isArrowFunctionExpression(transform)) {
        errors.push({
          message: `"callApi().then()" expects an arrow function, but got ${transform.type}`,
          node: transform,
          severity: "error",
        });
        return false;
      }
      if (!t.isExpression(transform.body)) {
        errors.push({
          message: `"callApi().then()" function body expects an expression, but got ${transform.body.type}`,
          node: transform.body,
          severity: "error",
        });
        return false;
      }
      if (transform.params.length !== 1) {
        errors.push({
          message: `"callApi().then()" function expects exactly one parameter, but got ${transform.params.length}`,
          node: transform.params[1] ?? transform,
          severity: "error",
        });
        return false;
      }
      const arg = transform.params[0];
      if (!t.isIdentifier(arg)) {
        errors.push({
          message: `"callApi().then()" function parameter expects an identifier, but got ${arg.type}`,
          node: arg,
          severity: "error",
        });
        return false;
      }
      const expr = `<% ${source.substring(transform.body.start!, transform.body.end!)} %>`;
      dataSource.transform = replaceVariables(
        expr,
        new Map([[arg.name, "DATA"]])
      );
    }

    dataSources.push(dataSource);
  }

  const body = ast?.program.body ?? [];

  if (ast) {
    t.traverse(ast.program, {
      enter(node) {
        if (t.isTSAnyKeyword(node)) {
          errors.push({
            message: `Unexpected 'any' type`,
            node,
            severity: "warning",
          });
        } else if (t.isTSAsExpression(node)) {
          errors.push({
            message: `Unexpected 'as' usage`,
            node,
            severity: "warning",
          });
        }
      },
    });
  }

  for (const stmt of body) {
    if (t.isVariableDeclaration(stmt)) {
      for (const dec of stmt.declarations) {
        if (t.isIdentifier(dec.id)) {
          contexts.push(dec.id.name);
        }
      }
    }
  }

  let exported = false;
  for (const stmt of body) {
    if (t.isVariableDeclaration(stmt)) {
      if (exported) {
        errors.push({
          message: `Unexpected variable declaration after export`,
          node: stmt,
          severity: "error",
        });
      }
      if (stmt.kind !== "let" && stmt.kind !== "const") {
        errors.push({
          message: `Unsupported variable declaration kind: ${stmt.kind}`,
          node: stmt,
          severity: "error",
        });
        continue;
      }
      if (stmt.declarations.length !== 1) {
        errors.push({
          message: `Expect exactly 1 declaration in a variable declaration statement, received ${stmt.declarations.length}`,
          node: stmt,
          severity: "error",
        });
        continue;
      }
      const dec = stmt.declarations[0];
      if (!t.isIdentifier(dec.id)) {
        errors.push({
          message: `Expect an identifier as the variable name, received ${dec.id.type}`,
          node: dec.id,
          severity: "error",
        });
        continue;
      }

      const name = dec.id.name;
      if (dec.init == null) {
        if (!dec.id.typeAnnotation) {
          errors.push({
            message: `Variable "${name}" with no initial value must have a type annotation`,
            node: dec.id,
            severity: "error",
          });
        }

        variables.push({ name });
      } else if (t.isAwaitExpression(dec.init)) {
        const call = dec.init.argument;
        if (!t.isCallExpression(call)) {
          errors.push({
            message: `Await expression must be a call expression, received ${call.type}`,
            node: call,
            severity: "error",
          });
          continue;
        }
        const { callee } = call;
        if (!(t.isIdentifier(callee) || t.isMemberExpression(callee))) {
          errors.push({
            message: `Await expression must call an identifier or member expression, received ${callee.type}`,
            node: callee,
            severity: "error",
          });
          continue;
        }
        if (t.isMemberExpression(callee)) {
          if (
            !t.isIdentifier(callee.property) ||
            callee.computed ||
            callee.property.name !== "then"
          ) {
            errors.push({
              message: `Unexpected awaited expression`,
              node: callee.property,
              severity: "error",
            });
          }
          parseDataSourceCall(callee.object, name, call.arguments);
        } else {
          parseDataSourceCall(call, name);
        }
      } else {
        const value = constructJsValue(dec.init, result, {
          allowExpression: true,
        });
        variables.push({ name, value });
      }
    } else if (t.isExportDefaultDeclaration(stmt)) {
      exported = true;
      const declaration = stmt.declaration;
      if (!t.isJSXElement(declaration)) {
        errors.push({
          message: `Exported default declaration must be a JSX element, but got ${declaration.type}`,
          node: declaration,
          severity: "error",
        });
        continue;
      }

      constructTsxView(declaration, result, options);
    } else {
      errors.push({
        message: `Unsupported top level statement type: ${stmt.type}`,
        node: stmt,
        severity: "error",
      });
      continue;
    }
  }

  return result;
}
