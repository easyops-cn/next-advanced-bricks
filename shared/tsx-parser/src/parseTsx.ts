import { parse, type ParseResult as BabelParseResult } from "@babel/parser";
import * as t from "@babel/types";
import type { StoryboardFunction } from "@next-core/types";
import type {
  Component,
  ParseResult,
  DataSource,
  ParseError,
  ParseOptions,
  Variable,
  Template,
} from "./interfaces.js";
import { parseLegacy } from "./parseLegacy.js";
import { parseSolid } from "./parseSolid.js";

export function parseTsx(source: string, options?: ParseOptions): ParseResult {
  const dataSources: DataSource[] = [];
  const variables: Variable[] = [];
  const components: Component[] = [];
  const errors: ParseError[] = [];
  let title: string | undefined;
  const componentsMap = new Map<string, Component>();
  const contexts: string[] = options?.withContexts ?? [];
  const contextGetters: string[] = [];
  const contextSetters = new Map<string, string>();
  const functions: StoryboardFunction[] = [];
  const functionNames: string[] = [];
  const contracts = new Set<string>();
  const templates: Template[] = [];
  const result: ParseResult = {
    source,
    title,
    dataSources,
    variables,
    components,
    componentsMap,
    errors,
    contexts,
    contextGetters,
    contextSetters,
    functionNames,
    functions,
    contracts,
    templates,
  };

  let ast: BabelParseResult<t.File> | undefined;
  try {
    ast = parse(source, {
      plugins: ["jsx", "typescript"],
      sourceType: "module",
      errorRecovery: true,
    });
  } catch (error) {
    errors.push({
      message: `Failed to parse TSX: ${error}`,
      node: null,
      severity: "fatal",
    });

    return result;
  }

  if (ast.errors?.length) {
    for (const error of ast.errors) {
      errors.push({
        message: `${error.code}: ${error.reasonCode}`,
        node: null,
        severity: "error",
      });
    }
  }

  for (const stmt of ast.program.body) {
    if (t.isExportDefaultDeclaration(stmt)) {
      if (t.isFunctionDeclaration(stmt.declaration)) {
        parseSolid(ast, result, options);
        return result;
      }
      break;
    }
  }

  parseLegacy(ast, result, options);
  return result;
}
