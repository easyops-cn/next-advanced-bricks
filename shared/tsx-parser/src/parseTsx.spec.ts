import { readFile } from "node:fs/promises";
import path from "node:path";
import { parseTsx } from "./parseTsx.js";

describe("parseTsx", () => {
  test("should work", async () => {
    const code = await readFile(
      path.join(__dirname, "./__fixtures__/logs.txt"),
      "utf-8"
    );

    const { source, ...result } = parseTsx(code);
    expect(result).toMatchSnapshot();
  });
});
