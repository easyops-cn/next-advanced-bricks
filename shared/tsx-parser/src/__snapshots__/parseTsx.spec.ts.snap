// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`parseTsx should work 1`] = `
{
  "components": [
    {
      "children": [
        {
          "children": [
            {
              "children": undefined,
              "componentId": undefined,
              "events": {
                "search": [
                  {
                    "action": "conditional",
                    "payload": {
                      "alternate": [
                        {
                          "action": "update_variable",
                          "payload": {
                            "name": "logId",
                            "value": undefined,
                          },
                        },
                        {
                          "action": "update_query",
                          "payload": {
                            "args": [
                              {
                                "logId": null,
                              },
                              {
                                "notify": false,
                              },
                            ],
                            "method": "push",
                          },
                        },
                      ],
                      "consequent": [
                        {
                          "action": "update_variable",
                          "payload": {
                            "name": "logId",
                            "value": "<% EVENT.detail %>",
                          },
                        },
                        {
                          "action": "update_query",
                          "payload": {
                            "args": [
                              {
                                "logId": "<% EVENT.detail %>",
                              },
                              {
                                "notify": false,
                              },
                            ],
                            "method": "push",
                          },
                        },
                      ],
                      "test": "<% EVENT.detail %>",
                    },
                  },
                ],
              },
              "name": "Search",
              "properties": {
                "inputStyle": {
                  "width": "300px",
                },
                "placeholder": "输入任务 ID 并按回车",
                "value": "<%= CTX.logId %>",
              },
              "ref": undefined,
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "name": "Plaintext",
                      "properties": {
                        "textContent": "<%= CTX.allExpanded ? "全部收起" : "全部展开" %>",
                      },
                    },
                  ],
                  "componentId": undefined,
                  "events": {
                    "click": [
                      {
                        "action": "update_variable",
                        "payload": {
                          "name": "allExpanded",
                          "value": "<% !CTX.allExpanded %>",
                        },
                      },
                    ],
                  },
                  "name": "Button",
                  "properties": {},
                  "ref": undefined,
                },
              ],
              "name": "If",
              "properties": {
                "dataSource": "<%= CTX.logs %>",
              },
            },
          ],
          "componentId": undefined,
          "events": undefined,
          "name": "Toolbar",
          "properties": {},
          "ref": undefined,
        },
        {
          "children": [
            {
              "children": undefined,
              "componentId": undefined,
              "events": undefined,
              "name": "Table",
              "properties": {
                "columns": [
                  {
                    "dataIndex": "time",
                    "key": "time",
                    "render": {
                      "children": [
                        {
                          "name": "Plaintext",
                          "properties": {
                            "textContent": "<%= CTX.__builtin_fn_mergeTexts({type:"expression",value:(moment(v * 1000).format("HH:mm:ss"))}) %>",
                          },
                        },
                      ],
                      "params": [
                        "v",
                      ],
                    },
                    "title": "时间",
                  },
                  {
                    "dataIndex": "opType",
                    "key": "opType",
                    "render": {
                      "children": [
                        {
                          "name": "Plaintext",
                          "properties": {
                            "textContent": "<%= CTX.__builtin_fn_mergeTexts({type:"expression",value:(\`\${v}\${r.opSubType ? \` / \${r.opSubType}\` : ""}\`)}) %>",
                          },
                        },
                      ],
                      "params": [
                        "v",
                        "r",
                      ],
                    },
                    "title": "操作类型",
                  },
                  {
                    "dataIndex": "data",
                    "key": "data",
                    "render": {
                      "children": [
                        {
                          "children": [
                            {
                              "children": undefined,
                              "componentId": undefined,
                              "events": undefined,
                              "name": "CodeBlock",
                              "properties": {
                                "language": "json",
                                "source": "<%= FN.prettierOutput(v) %>",
                                "style": {
                                  "maxWidth": "696px",
                                },
                              },
                              "ref": undefined,
                            },
                            {
                              "children": [
                                {
                                  "name": "Plaintext",
                                  "properties": {
                                    "textContent": "<%= v %>",
                                  },
                                },
                              ],
                              "componentId": undefined,
                              "events": undefined,
                              "name": "Plaintext",
                              "properties": {
                                "style": {
                                  "display": "block",
                                  "maxWidth": "696px",
                                  "overflow": "hidden",
                                  "textOverflow": "ellipsis",
                                  "whiteSpace": "nowrap",
                                },
                              },
                              "ref": undefined,
                              "slot": "else",
                            },
                          ],
                          "name": "If",
                          "properties": {
                            "dataSource": "<%= CTX.allExpanded || CTX.expandedSet.has(r.logId) %>",
                          },
                        },
                      ],
                      "params": [
                        "v",
                        "r",
                      ],
                    },
                    "title": "数据",
                  },
                  {
                    "dataIndex": "operation",
                    "key": "operation",
                    "render": {
                      "children": [
                        {
                          "children": undefined,
                          "componentId": undefined,
                          "events": {
                            "open.detail": [
                              {
                                "action": "update_variable",
                                "payload": {
                                  "name": "detailLog",
                                  "value": "<% r %>",
                                },
                              },
                              {
                                "action": "call_ref",
                                "payload": {
                                  "args": [],
                                  "method": "open",
                                  "ref": "modalRef",
                                  "scope": "view",
                                },
                              },
                            ],
                            "toggle": [
                              {
                                "action": "update_variable",
                                "payload": {
                                  "name": "expandedSet",
                                  "value": "<% new Set(
                        CTX.expandedSet.has(r.logId)
                          ? [...CTX.expandedSet].filter((id) => id !== r.logId)
                          : [...CTX.expandedSet, r.logId]
                      ) %>",
                                },
                              },
                            ],
                          },
                          "name": "Operations",
                          "properties": {
                            "allExpanded": "<%= CTX.allExpanded %>",
                            "expanded": "<%= CTX.expandedSet.has(r.logId) %>",
                          },
                          "ref": undefined,
                        },
                      ],
                      "params": [
                        "v",
                        "r",
                      ],
                    },
                    "title": "操作",
                  },
                ],
                "dataSource": {
                  "list": "<%= CTX.logs %>",
                },
                "pagination": false,
                "rowKey": "logId",
              },
              "ref": undefined,
            },
          ],
          "name": "If",
          "properties": {
            "dataSource": "<%= CTX.logs %>",
          },
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": undefined,
                  "componentId": undefined,
                  "events": undefined,
                  "name": "Descriptions",
                  "properties": {
                    "columns": 1,
                    "dataSource": "<%= CTX.detailLog %>",
                    "list": [
                      {
                        "field": "logId",
                        "label": "日志 ID",
                      },
                      {
                        "field": "conversationId",
                        "label": "任务 ID",
                      },
                      {
                        "field": "taskId",
                        "label": "任务 ID",
                      },
                      {
                        "field": "time",
                        "label": "时间",
                        "render": {
                          "children": [
                            {
                              "name": "Plaintext",
                              "properties": {
                                "textContent": "<%= CTX.__builtin_fn_mergeTexts({type:"expression",value:(moment(r.time * 1000).format("llll"))}) %>",
                              },
                            },
                          ],
                          "params": [
                            "r",
                          ],
                        },
                      },
                    ],
                  },
                  "ref": undefined,
                },
                {
                  "children": undefined,
                  "componentId": undefined,
                  "events": undefined,
                  "name": "CodeBlock",
                  "properties": {
                    "language": "json",
                    "source": "<%= FN.prettierOutput(CTX.detailLog.data) %>",
                  },
                  "ref": undefined,
                },
              ],
              "name": "If",
              "properties": {
                "dataSource": "<%= CTX.detailLog %>",
              },
            },
          ],
          "componentId": undefined,
          "events": undefined,
          "name": "Modal",
          "properties": {
            "hideCancelButton": true,
            "maskClosable": true,
            "title": "<%= CTX.detailLog
            ? \`\${r.opType}\${r.opSubType ? \` / \${r.opSubType}\` : ""}\`
            : "" %>",
            "width": "800px",
          },
          "ref": "modalRef",
        },
      ],
      "componentId": undefined,
      "events": undefined,
      "name": "View",
      "properties": {
        "title": "任务日志",
      },
      "ref": undefined,
    },
  ],
  "componentsMap": Map {},
  "contextSetters": Map {
    "setLogId" => "logId",
    "setAllExpanded" => "allExpanded",
    "setExpandedSet" => "expandedSet",
    "setDetailLog" => "detailLog",
  },
  "contexts": [
    "logId",
    "allExpanded",
    "expandedSet",
    "detailLog",
    "logs",
  ],
  "contracts": Set {
    "<% \`logs/\${CTX.logId}\` %>",
  },
  "dataSources": [
    {
      "api": "<% \`logs/\${CTX.logId}\` %>",
      "config": {
        "enabled": "<% !!CTX.logId %>",
        "fallback": null,
      },
      "http": true,
      "name": "logs",
      "scope": "view",
      "transform": "<% DATA.data.logs %>",
    },
  ],
  "errors": [],
  "functionNames": [
    "prettierOutput",
  ],
  "functions": [
    {
      "name": "prettierOutput",
      "source": "function prettierOutput(output: string) {
  try {
    const data = JSON.parse(output);
    return JSON.stringify(data, null, 2);
  } catch (e) {
    return output;
  }
}",
      "typescript": true,
    },
  ],
  "globals": Map {
    "query" => "QUERY",
  },
  "refs": [
    "modalRef",
  ],
  "templates": [
    {
      "components": [
        {
          "children": [
            {
              "children": [
                {
                  "children": undefined,
                  "componentId": undefined,
                  "events": {
                    "click": [
                      {
                        "action": "dispatch_event",
                        "payload": {
                          "detail": "<% EVENT.detail %>",
                          "type": "toggle",
                        },
                      },
                    ],
                  },
                  "name": "Button",
                  "properties": {
                    "textContent": "<%= CTX.__builtin_fn_mergeTexts({type:"expression",value:(STATE.expanded ? "收起" : "展开")}) %>",
                  },
                  "ref": undefined,
                },
              ],
              "name": "If",
              "properties": {
                "dataSource": "<%= !STATE.allExpanded %>",
              },
            },
            {
              "children": undefined,
              "componentId": undefined,
              "events": {
                "click": [
                  {
                    "action": "dispatch_event",
                    "payload": {
                      "detail": "<% EVENT.detail %>",
                      "type": "open.detail",
                    },
                  },
                ],
              },
              "name": "Button",
              "properties": {
                "textContent": "详情",
              },
              "ref": undefined,
            },
          ],
          "componentId": undefined,
          "events": undefined,
          "name": "Toolbar",
          "properties": {},
          "ref": undefined,
        },
      ],
      "dataSources": [],
      "name": "Operations",
      "variables": [
        {
          "expose": true,
          "name": "expanded",
        },
        {
          "expose": true,
          "name": "allExpanded",
        },
      ],
    },
  ],
  "variables": [
    {
      "expose": false,
      "name": "logId",
      "value": "<% QUERY.logId %>",
    },
    {
      "expose": false,
      "name": "allExpanded",
      "value": false,
    },
    {
      "expose": false,
      "name": "expandedSet",
      "value": "<% new Set() %>",
    },
    {
      "expose": false,
      "name": "detailLog",
      "value": null,
    },
  ],
}
`;
